pkg/mimirpb/prealloc_rw2.go

```diff
@@ -30,27 +34,50 @@ func FromWriteRequestToRW2Request(rw1 *WriteRequest, commonSymbols []string, off
 	defer reuseSymbolsTable(symbols)
 	symbols.ConfigureCommonSymbols(offset, commonSymbols)
 
+	var metadataMap map[string]*MetricMetadata
+	if len(rw1.Metadata) > 0 {
```

Comment:
Throughout we take extra care to never allocate or interact with the map if the request has no metadata.
This keeps the impact on regular series pushes low.